checkpoint/handler_test.go

```diff
@@ -1,331 +1,406 @@
+	suite.handler = checkpoint.NewHandler(suite.app.CheckpointKeeper, &suite.contractCaller)
+	suite.sideHandler = checkpoint.NewSideTxHandler(suite.app.CheckpointKeeper, &suite.contractCaller)
+	suite.postHandler = checkpoint.NewPostTxHandler(suite.app.CheckpointKeeper, &suite.contractCaller)
+}
 
-	bankTypes.RegisterCodec(cdc)
+func TestHandlerTestSuite(t *testing.T) {
+	suite.Run(t, new(HandlerTestSuite))
+}
 
-	// custom types
-	borTypes.RegisterCodec(cdc)
-	checkpointTypes.RegisterCodec(cdc)
-	stakingTypes.RegisterCodec(cdc)
+func (suite *HandlerTestSuite) TestHandler() {
+	t, ctx := suite.T(), suite.ctx
 
-	cdc.Seal()
-	return cdc
+	// side handler
+	result := suite.handler(ctx, nil)
+	require.False(t, result.IsOK(), "Handler should fail")
 }
 
-// init for test cases
-func CreateTestInput(t *testing.T, isCheckTx bool) (sdk.Context, staking.Keeper, Keeper) {
-	helper.InitHeimdallConfig(os.ExpandEnv("$HOME/.heimdalld"))
-	db := dbm.NewMemDB()
-	ms := store.NewCommitMultiStore(db)
-
-	// TODO create more keys like borKey etc
-	keyCheckpoint := sdk.NewKVStoreKey("checkpoint")
-	keyStaking := sdk.NewKVStoreKey("staking")
-	keyMaster := sdk.NewKVStoreKey("master")
-	keyParams := sdk.NewKVStoreKey(subspace.StoreKey)
-	tKeyParams := sdk.NewTransientStoreKey(subspace.TStoreKey)
-
-	// mount all
-	ms.MountStoreWithDB(keyCheckpoint, sdk.StoreTypeIAVL, db)
-	ms.MountStoreWithDB(keyStaking, sdk.StoreTypeIAVL, db)
-	ms.MountStoreWithDB(keyMaster, sdk.StoreTypeIAVL, db)
-	ms.MountStoreWithDB(keyParams, sdk.StoreTypeIAVL, db)
-	ms.MountStoreWithDB(tKeyParams, sdk.StoreTypeTransient, db)
-
-	err := ms.LoadLatestVersion()
-	require.Nil(t, err)
-
-	ctx := sdk.NewContext(ms, abci.Header{ChainID: "foochainid"}, isCheckTx, log.NewNopLogger())
-	cdc := MakeTestCodec()
-
-	paramsKeeper := params.NewKeeper(cdc, keyParams, tKeyParams, common.DefaultCodespace)
-
-	dummyStakingKeeper := staking.Keeper{}
-
-	checkpointKeeper := NewKeeper(
-		cdc,
-		keyCheckpoint,
-		paramsKeeper.Subspace(checkpointTypes.DefaultParamspace),
-		common.DefaultCodespace,
-		dummyStakingKeeper,
-	)
+// test handler for message
+func (suite *HandlerTestSuite) TestHandleMsgCheckpoint() {
+	t, app, ctx := suite.T(), suite.app, suite.ctx
+	keeper := app.CheckpointKeeper
+	stakingKeeper := app.StakingKeeper
+	topupKeeper := app.TopupKeeper
+	start := uint64(0)
+	maxSize := uint64(256)
+	borChainId := "1234"
+	params := keeper.GetParams(ctx)
+	dividendAccount := hmTypes.DividendAccount{
+		User:      hmTypes.HexToHeimdallAddress("123"),
+		FeeAmount: big.NewInt(0).String(),
+	}
+	topupKeeper.AddDividendAccount(ctx, dividendAccount)
 
-	stakingKeeper := staking.NewKeeper(
-		cdc,
-		keyStaking,
-		paramsKeeper.Subspace(stakingTypes.DefaultParamspace),
-		common.DefaultCodespace,
-		checkpointKeeper,
-	)
-	checkpointKeeper.sk = stakingKeeper
+	// check valid checkpoint
+	// generate proposer for validator set
+	cmn.LoadValidatorSet(2, t, stakingKeeper, ctx, false, 10)
```

Comment:
We should move function to `staking/simulation` package since this is very specific to staking and remove `github.com/maticnetwork/heimdall/test` folder.